### PR TITLE
[CALCITE-2376] Unify ES2 and ES5 adapters. Migrate to low-level ES rest client

### PR DESCRIPTION
**Pull request by asereda-gs**
## Major change to elastic search adapter(s)

 1. Combine ES2 and ES5 under single maven module (called `elasticsearch`). Move shared ES source inside core to   elasticsearch module.
2. Migrate transport layer to [low-level rest client](https://www.elastic.co/guide/en/elasticsearch/client/java-rest/current/java-rest-low.html). There are no more heavy core server-side ES dependencies (like Lucene, Netty etc.) in adapter (just apache http client and apache codecs).
    Native Transport client will be removed by the vendor in the next release (ES7-8).
3. Manually parse search results using jackson JSON library (this is already a transitive dependency from core)

## Expected breaking changes for clients:
1. Change class names of Schema and SchemaFactory. They don't contain ES major version number anymore (2, 5, 6 etc.):
* ElasticsearchSchemaFactory(new) vs Elasticsearch{2|5}SchemaFactory (old)
* ElasticsearchSchema(new) vs Elasticsearch{2|5}Schema (old)
* changed package name(s)
This affects existing configurations (especially JSON files)

2. There is a single maven module `elasticsearch`. `elasticsearch2` and `elasticsearch5` are removed.
    Client will have to manually update their artifact dependencies.

### CODE REVIEW

File: elasticsearch/src/main/java/org/apache/calcite/adapter/elasticsearch/ElasticsearchSchemaFactory.java [hunk 1]
```diff
@@ -49,15 +58,37 @@ public Elasticsearch5SchemaFactory() {
...
               new TypeReference<Map<String, Integer>>() { });
+
+      final RestClient client = connect(coordinates);
+
       final Map<String, String> userConfig =
           mapper.readValue((String) map.get("userConfig"),
               new TypeReference<Map<String, String>>() { });
+
       final String index = (String) map.get("index");
-      return new Elasticsearch5Schema(coordinates, userConfig, index);
```
Comment (beikov) [inline elasticsearch/src/main/java/org/apache/calcite/adapter/elasticsearch/ElasticsearchSchemaFactory.java]: What's with the userConfig map? Before, a user was able to specify driver settings through this like e.g. `bulk.flush.max.actions`, `bulk.flush.max.size.mb` and `bulk.flush.interval.ms`. Are there any settings that we could allow users to do through this?

File: elasticsearch/src/main/java/org/apache/calcite/adapter/elasticsearch/ElasticsearchTable.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+import java.util.Locale;
+import java.util.Map;
+
+/**
+ * Table based on an Elasticsearch5 type.
+ */
+public class ElasticsearchTable extends AbstractElasticsearchTable {
+  private final RestClient restClient;
+  private final ElasticsearchVersion version;
+  private final ObjectMapper mapper;
```
Comment (eolivelli) [inline elasticsearch/src/main/java/org/apache/calcite/adapter/elasticsearch/ElasticsearchTable.java]: I think ObjectMapper can be used as static final field. It has no state and it caches metadata so it usually it is a good deal

File: elasticsearch/src/main/java/org/apache/calcite/adapter/elasticsearch/ElasticsearchSchemaFactory.java (hunk 1 above)
Comment (asereda-gs) [reply elasticsearch/src/main/java/org/apache/calcite/adapter/elasticsearch/ElasticsearchSchemaFactory.java]: Those changes were applicable only for legacy transport client. New low-level rest client is [configured programatically](https://www.elastic.co/guide/en/elasticsearch/client/java-rest/current/_common_configuration.html). We may add some parameters in future (like timeout).

File: elasticsearch/src/main/java/org/apache/calcite/adapter/elasticsearch/ElasticsearchTable.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+  /**
+   * Creates an Elasticsearch5Table.
```
[the remark this answers is not in the document]
Comment (asereda-gs) [reply elasticsearch/src/main/java/org/apache/calcite/adapter/elasticsearch/ElasticsearchTable.java]: Done

**Comment by michaelmior** [comment]
No major additional concerns here although we may want to hold off on merging this until after the next release. I'm not sure how heavily the ES adapters are used but this is a fairly big change and I wouldn't want to rush it.

**Comment by asereda-gs** [comment]
@beikov regarding visibility. Before there were 3 ES-related packages:
1. elasticsearch (in `core` module)
2. elasticsearch2 (in `elasticsearch2 ` module)
3. elasticsearch5 (in `elasticsearch5 ` module)

Base classes located in `core` were mostly public / protected (so they could be extended in other packages). Now, all ES related classes reside in the same package so visibility can be reduced to package-private. 

Philosophically speaking I just prefer to minimize accessibility of classes and members and provide explicit access points. Should one field / method be required by clients, one can always increase the visibility, it is much harder to do the opposite.

**Comment by michaelmior** [comment]
@asereda-gs If it's occurring with multiple adapters, I assume it's unrelated to these changes, so I wouldn't be concerned about that. That would be a good bug to be fixed for the upcoming release if you can put a PR together.

**Comment by asereda-gs** [comment]
@michaelmior Yes it's occurring with, at least, ES and Mongo.  But if I'm making a patch for [CALCITE-2331](https://issues.apache.org/jira/browse/CALCITE-2331) I'd rather do it against single branch. For instance, ES Query generation and execution depend on transport client and ES API being used.

**Comment by michaelmior** [comment]
@asereda-gs I would suggest making a patch against the branch for this PR. Once that's confirmed to work, we can try it on master as well and hopefully it works there and we can merge it independently. If not, we can discuss the best way to get everything in.

**Comment by asereda-gs** [comment]
@michaelmior  I have created #753 (patch for (A or B) and C) based on this branch.
Let's discuss it.

**Comment by michaelmior** [comment]
Unfortunately due to some major changes that just landed on master, this PR will need to updated. Although you'll see a lot of conflicts, hopefully this should not be too painful to resolve.

**Comment by asereda-gs** [comment]
@michaelmior Conflicts have been resolved

**Comment by michaelmior** [comment]
@beikov Any other thoughts on this? As mentioned before, I'm not an ES user, so it would be good to have a second look from someone who is before merging.

**Pull request closed without merging**